### ISSUE TITLE
Calling `emtrends` from within a function

### ISSUE THREAD

**Issue by cimentadaj** [rvlenth/emmeans#153]
This error is the same as https://github.com/rvlenth/emmeans/issues/13. 

See the example below:

```r
library(emmeans)
library(lme4)
fm1 <- lmer(mpg ~ drat*carb + (1 | cyl), mtcars)

inside_fun <- function(model) {
  emtrends(
    model,
    "carb",
    "drat"
  )
}

inside_fun(fm1)
#> Warning in as.vector(x): restarting interrupted promise evaluation
#> Error in as.vector(x) : object 'model' not found
#> Error in ref_grid(object = model, options = list(just.data = TRUE)): Perhaps a 'data' or 'params' argument is needed
```

I've look at the source code and figured out where the problem is. Inside `emtrends` you deconstruct the matched call to add `ref_grid` and then eval it. The problem is you're evaluating all arguments from the calling environment. This is fine when you're using the function interactively (which throws no errors) but when called from a function, it can't find the model object.

```r
library(emmeans)
library(lme4)
fm1 <- lmer(mpg ~ drat*carb + (1 | cyl), mtcars)

# Unfixed emtrends
emtrends_dev <- function (object, specs, var, delta.var = 0.001 * rng, max.degree = 1, ...) {
  estName = paste(var, "trend", sep = ".")
  cl = match.call()
  cl[[1]] = quote(ref_grid)
  cl$var = cl$specs = NULL
  if (is.null(cl$options)) 
    cl$options = list()
  if (missing(object) && ("model" %in% names(cl))) {
    names(cl)[names(cl) == "model"] = "object"
    object = eval(cl$object)
  }
  cl$options$just.data = TRUE
  # The error is here: evaluating cl in the parent call and can't find the model object.
  data = eval(cl)
  # I delete the remaining long part of the function
  data
}

inside_fun <- function(model) {
  emtrends_dev(
    model,
    "carb",
    "drat"
  )
}

inside_fun(fm1)
#> ref_grid(object = model, options = list(just.data = TRUE))
#> Warning in as.vector(x): restarting interrupted promise evaluation
#> Error in as.vector(x) : object 'model' not found
#> Error in ref_grid(object = model, options = list(just.data = TRUE)): Perhaps a 'data' or 'params' argument is needed

# Fixed emtrends
emtrends_dev <- function (object, specs, var, delta.var = 0.001 * rng, max.degree = 1, ...) {
  estName = paste(var, "trend", sep = ".")
  cl = match.call()
  cl[[1]] = quote(ref_grid)
  cl$var = cl$specs = NULL
  if (is.null(cl$options)) 
    cl$options = list()
  if (missing(object) && ("model" %in% names(cl))) {
    names(cl)[names(cl) == "model"] = "object"
    object = eval(cl$object)
  }
  cl$options$just.data = TRUE

  # Here I add the quoted object model (which is defined in the function's environment)
  # and eval it from the current environment
  cl$object <- quote(object)
  data = eval(cl, envir = environment())
  data
}

# This fixes it for calling it from within functions
inside_fun(fm1)
#>                     drat carb
#> Mazda RX4           3.90    4
#> Mazda RX4 Wag       3.90    4
#> Datsun 710          3.85    1
#> Hornet 4 Drive      3.08    1
#> Hornet Sportabout   3.15    2
#> Valiant             2.76    1
#> Duster 360          3.21    4
#> Merc 240D           3.69    2
#> Merc 230            3.92    2
#> Merc 280            3.92    4
#> Merc 280C           3.92    4
#> Merc 450SE          3.07    3
#> Merc 450SL          3.07    3
#> Merc 450SLC         3.07    3
#> Cadillac Fleetwood  2.93    4
#> Lincoln Continental 3.00    4
#> Chrysler Imperial   3.23    4
#> Fiat 128            4.08    1
#> Honda Civic         4.93    2
#> Toyota Corolla      4.22    1
#> Toyota Corona       3.70    1
#> Dodge Challenger    2.76    2
#> AMC Javelin         3.15    2
#> Camaro Z28          3.73    4
#> Pontiac Firebird    3.08    2
#> Fiat X1-9           4.08    1
#> Porsche 914-2       4.43    2
#> Lotus Europa        3.77    2
#> Ford Pantera L      4.22    4
#> Ferrari Dino        3.62    6
#> Maserati Bora       3.54    8
#> Volvo 142E          4.11    2

# And also for using it interactively.
emtrends_dev(
  fm1,
  "carb",
  "drat"
)
#>                     drat carb
#> Mazda RX4           3.90    4
#> Mazda RX4 Wag       3.90    4
#> Datsun 710          3.85    1
#> Hornet 4 Drive      3.08    1
#> Hornet Sportabout   3.15    2
#> Valiant             2.76    1
#> Duster 360          3.21    4
#> Merc 240D           3.69    2
#> Merc 230            3.92    2
#> Merc 280            3.92    4
#> Merc 280C           3.92    4
#> Merc 450SE          3.07    3
#> Merc 450SL          3.07    3
#> Merc 450SLC         3.07    3
#> Cadillac Fleetwood  2.93    4
#> Lincoln Continental 3.00    4
#> Chrysler Imperial   3.23    4
#> Fiat 128            4.08    1
#> Honda Civic         4.93    2
#> Toyota Corolla      4.22    1
#> Toyota Corona       3.70    1
#> Dodge Challenger    2.76    2
#> AMC Javelin         3.15    2
#> Camaro Z28          3.73    4
#> Pontiac Firebird    3.08    2
#> Fiat X1-9           4.08    1
#> Porsche 914-2       4.43    2
#> Lotus Europa        3.77    2
#> Ford Pantera L      4.22    4
#> Ferrari Dino        3.62    6
#> Maserati Bora       3.54    8
#> Volvo 142E          4.11    2
```

There's another `eval` call from within the function that would need to change `envir` to `environment()` but I think that should be it. If you're open to it, I could send a pull request with these small changes for you to review.

**Comment by cimentadaj** [comment]
To make it easier for you to evaluate, here's the **fixed** function with comments where I've changed stuff.

```r
emtrends_dev <- function (object, specs, var, delta.var = 0.001 * rng, max.degree = 1, ...) {
  estName = paste(var, "trend", sep = ".")
  cl = match.call()
  cl[[1]] = quote(ref_grid)
  cl$var = cl$specs = NULL
  if (is.null(cl$options)) 
    cl$options = list()
  if (missing(object) && ("model" %in% names(cl))) {
    names(cl)[names(cl) == "model"] = "object"
    object = eval(cl$object)
  }
  cl$options$just.data = TRUE
  # Adds the quoted model in the matched call
  # and evaluate it in the current function environment
  cl$object <- quote(object)
  data = eval(cl, envir = environment())
  cl$options$just.data = NULL
  x = data[[var]]
  fcn = NULL
  if (is.null(x)) {
    fcn = var
    var = .all.vars(as.formula(paste("~", var)))
    if (length(var) > 1) 
      stop("Can only support a function of one variable")
    else {
      x = data[[var]]
      if (is.null(x)) 
        stop("Variable '", var, "' is not in the dataset")
    }
  }
  rng = diff(range(x))
  if (delta.var == 0) 
    stop("Provide a nonzero value of 'delta.var'")
  max.degree = max(1, min(5, as.integer(max.degree + 0.1)))
  delts = delta.var * (0:max.degree)
  idx.base = as.integer((2 + max.degree)/2)
  delts = delts - delts[idx.base]
  cl$data = quote(data)
  cl$options$var = var
  cl$options$delts = delts

  bigRG = eval(cl, envir = environment())
  var.subs = as.list(as.data.frame(matrix(seq_len(nrow(bigRG@grid)), 
                                          ncol = length(delts))))
  RG = orig.rg = bigRG[var.subs[[idx.base]]]
  row.names(RG@grid) = seq_along(RG@grid[[1]])
  linfct = lapply(seq_along(delts), function(i) bigRG@linfct[var.subs[[i]], 
                                                           , drop = FALSE])
  if (!is.null(fcn)) {
    tmp = sapply(seq_along(delts), function(i) eval(parse(text = fcn), 
                                                    envir = bigRG@grid[var.subs[[i]], , drop = FALSE]))
    delta.var = apply(tmp, 1, function(.) mean(diff(.)))
  }
  newlf = numeric(0)
  h = 1
  for (i in 1:max.degree) {
    linfct = lapply(seq_along(linfct)[-1], function(j) linfct[[j]] - 
                                                         linfct[[j - 1]])
    h = h * delta.var * i
    what = as.integer((length(linfct) + 1)/2)
    newlf = rbind(newlf, linfct[[what]]/h)
  }
  RG@linfct = newlf
  RG@roles$trend = var
  if (max.degree > 1) {
    degnms = c("linear", "quadratic", "cubic", "quartic", 
               "quintic")
    RG@grid$degree = degnms[1]
    g = RG@grid
    for (j in 2:max.degree) {
      g$degree = degnms[j]
      RG@grid = rbind(RG@grid, g)
    }
    RG@roles$predictors = c(RG@roles$predictors, "degree")
    RG@levels$degree = degnms[1:max.degree]
  }
  RG@grid$.offset. = NULL
  RG@misc$tran = RG@misc$tran.mult = NULL
  RG@misc$estName = estName
  RG@misc$methDesc = "emtrends"
  emmeans:::.save.ref_grid(RG)
  if (missing(specs) || is.null(specs)) 
    return(RG)
  args = list(object = NULL, specs = specs, ...)
  args$at = args$cov.reduce = args$mult.levs = args$vcov. = args$data = args$trend = args$transform = NULL
  if (max.degree > 1) {
    chk = union(all.vars(specs), args$by)
    if (!("degree" %in% chk)) 
      args$by = c("degree", args$by)
  }
  args$object = RG
  do.call("emmeans", args)
}

```

**Comment by rvlenth** [comment]
Thanks. I really get confused with this scoping stuff.

That said, Your example (at least) works just fine if I add just one line of code:
```r
cl$object = quote(object)
```
I don't think I need to put `envir = environment()` in the `eval()` expressions. Can you think of an example where that too is needed?

**Comment by rvlenth** [comment]
PS -- if I set a breakpoint somewhere in `emmtrends()`, then with your example, I get:
```
Browse[2]> ls(parent.frame())
[1] "model"
```
So clearly, `model` can be found within the default environment for `eval()`.

**Comment by cimentadaj** [comment]
Well, then I also got slightly confused. In any case, it should be fixed with `quote`.

**Comment by rvlenth** [comment]
Version 1.4.3 is now on CRAN, fixes this issue (as well as #158 and #160)
